server/src/main/java/io/deephaven/server/console/ConsoleServiceGrpcImpl.java

```diff
@@ -303,57 +316,145 @@ public void onCompleted() {
+
+        @Override
+        public void run() {
+            while (true) {
+                if (!guard.compareAndSet(false, true)) {
+                    return;
+                }
+                boolean queueIsKnownEmpty;
+                try {
+                    queueIsKnownEmpty = sendInternal();
+                } finally {
+                    guard.set(false);
+                }
+                if (done) {
+                    return;
+                }
+                if (!client.isReady()) {
+                    // When !client.isReady(), onReady() is going to be called and will handle the reschedule.
+                    // Note: it's important that client.isReady() is checked _outside_ of the guard block (otherwise,
+                    // the onReady() call could execute and schedule before the current thread exits the guard block,
+                    // and we'd fail to successfully reschedule).
+                    return;
+                }
+                if (queueIsKnownEmpty) {
+                    scheduler.runAfterDelay(SUBSCRIBE_TO_LOGS_SEND_MILLIS, this);
+                    return;
+                }
+                // Continue sending until we are done, the client isn't ready, or we know the queue is empty.
+            }
+        }
 
-            // TODO this is not a good implementation, just a quick one, but it does appear to be safe,
-            // since LogBuffer is synchronized on access to the listeners. We're on the same thread
-            // as all other log receivers and
-            try {
-                LogSubscriptionData payload = LogSubscriptionData.newBuilder()
-                        .setMicros(record.getTimestampMicros())
-                        .setLogLevel(record.getLevel().getName())
-                        // this could be done on either side, doing it here because its a weird charset and we should
-                        // own that
-                        .setMessage(record.getDataString())
-                        .build();
-                synchronized (responseObserver) {
-                    responseObserver.onNext(payload);
+        private boolean sendInternal() {
+            while (!done && client.isReady()) {
+                LogSubscriptionData payload = dequeue();
+                if (payload == null) {
+                    // we know the queue is empty
+                    return true;
                 }
-            } catch (Throwable ignored) {
-                // we are ignoring exceptions here deliberately, and just shutting down
-                close();
+                GrpcUtil.safelyOnNext(client, payload);
+            }
+            // we don't know if the queue is empty
+            return false;
+        }
+
+        // ------------------------------------------------------------------------------------------------------------
+
+        private void onReady() {
+            scheduler.runImmediately(this);
```

Comment:
I think this should be a no-op if `guard.get() == true`.

Reply:
I _think_ you are right, but it's a _very_ rare case. In some testing I did, it only saved us from executing 

```
>>> 13/3583491
3.6277473558605283e-06
```

times.